### PR TITLE
Proofs of poly_frommsg

### LINKED ISSUE #336: CBMC: State + prove spec for poly_frommsg()

### PR DESCRIPTION
**Pull request by rod-chapman**
- Fixes #336 

Adds proof of poly_frommsg() function. In particular:

1. Strengthen post-condition of cmov_int16() to use OBJECT_UPTO rather than OBJECT_WHOLE to allow a caller to modify one element of a larger array.
2. Add contracts to specification and body of poly_frommsg()
3. Add CBMC proof artefacts for this function.

### CODE REVIEW

File: mlkem/poly.c [hunk 1]
```diff
@@ -274,28 +274,38 @@ void poly_frombytes(poly *r, const uint8_t a[MLKEM_POLYBYTES]) {
...
-    }
-  }
+  const int num_blocks = MLKEM_N / 8;
+
+  for (int i = 0; i < num_blocks; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_WHOLE(r))
+    INVARIANT(i >= 0)
+    INVARIANT(i <= num_blocks)
+    INVARIANT(num_blocks == 32)
```
Comment (hanno-becker) [inline mlkem/poly.c:289]: This should not be needed?

File: mlkem/poly.c [hunk 2]
```diff
@@ -274,28 +274,38 @@ void poly_frombytes(poly *r, const uint8_t a[MLKEM_POLYBYTES]) {
...
+        ASSIGNS(j, OBJECT_WHOLE(r))
+        INVARIANT(i >= 0 && i <  num_blocks)
+        INVARIANT(j >= 0 && j <= 8)
+        INVARIANT(ARRAY_IN_BOUNDS(int, k, 0, (8 * i + j - 1), r->coeffs, 0, (MLKEM_Q - 1)))
+        // clang-format on
+        {
+          r->coeffs[8 * i + j] = 0;
+          cmov_int16(&r->coeffs[8 * i + j], HALF_Q, (msg[i] >> j) & 1);
 
+          // Check that the value just assigned must be either HALF_Q or 0
```
Comment (hanno-becker) [inline mlkem/poly.c:304]: This can be removed? It is a verbatim restatement of the post condition of `cmov_int16`?

**Review by hanno-becker — CHANGES_REQUESTED** [review]
LGTM, but can you trim down on unneeded invariants and assertions please?

File: mlkem/poly.c (hunk 1 above)
Comment (rod-chapman) [reply mlkem/poly.c:289]: Removing that num_blocks == 32 invariants changes the time to prove this function (1 CPU on my laptop) from 3 min 47 seconds to 5 min 20 seconds. Worth removing or not?

File: mlkem/poly.c (hunk 2 above)
Comment (rod-chapman) [reply mlkem/poly.c:304]: Yes... removing it does not seem to change proof performance, so it can go.

File: mlkem/poly.c (hunk 1 above)
Comment (hanno-becker) [reply mlkem/poly.c:289]: Can we not use write `MLKEM_N / 8` everywhere? That's how it's done in the other loops.

Comment (hanno-becker) [reply mlkem/poly.c]: I took the liberty to change that.

File: mlkem/poly.c (hunk 2 above)
Comment (hanno-becker) [reply mlkem/poly.c]: I took the liberty to change that.

**Review by hanno-becker — APPROVED** [review]

File: mlkem/poly.c (hunk 1 above)
Comment (rod-chapman) [reply mlkem/poly.c]: OK. I personally find the repetition dubious, but happy to oblige. I guess the repetition is only risky if there's a chance the value might change, which is in this case is zero...

File: mlkem/poly.c (hunk 2 above)
Comment (rod-chapman) [reply mlkem/poly.c]: OK

**Pull request merged**